### DISCUSSION TITLE
How would you handle passwords (async hashing)?

### DISCUSSION THREAD

**Discussion by mikecann** [reimagined/resolve#1772] [Q&A]
Bcrypt generates passwords asyncronously. Command handlers need to be syncronous I believe. Obviously you dont want to put the password in plain text in an event so how does one handle this scenario?

**Comment by superroma** [comment]
I guess you should create a password hash before sending a command.
Here is user payload generation in Hacker News example - you would create a password hash from supplied password here:
https://github.com/reimagined/resolve/blob/960731601f017ff8da059ed8a2ee4646472f45fa/examples/hacker-news/auth/route_register_callback.js#L16-L19

**Reply by mikecann** [reply]
Ah cool, so there is a layer (like express or something) that sits infront of the command handlers that accepts the HTTP requests then passes the payload to the command handler. Its in that section that you do your async stuff such as bcrypting passwords?

**Reply by superroma** [reply]
Yes, we have a place for generic API handlers https://reimagined.github.io/resolve/docs/api-handlers

In this particular scenario there is a module-auth that provides these api handlers https://reimagined.github.io/resolve/docs/authentication-and-authorization

**Reply by mikecann** [reply]
> In this particular scenario there is a module-auth that provides these api handlers https://reimagined.github.io/resolve/docs/authentication-and-authorization

I got to admit I find these "modules" a little confusing. Are they basically plugins that can add pre-packaged set of functionality to your app?

**Reply by superroma** [reply]
Yes, module is just a reusable part of your app - it can contain anything. 
module-auth is essentially a package with a set of pre-configured routes and api handlers.

Another example is https://github.com/reimagined/resolve/tree/dev/packages/modules/module-comments - module, that contains aggregates, command handlers, read-models and even renderless react UI components.

I admit there is not much information how to author modules and use them. This is something that we going to elaborate in the future, and we have many reusable modules in mind - like authorisation, org structure memberships, permissions, roles, payments. But at the moment we are focused on the other things.
